### PR TITLE
GoodweAPI fixes

### PR DESCRIPTION
**Pull request by llagendijk**
hi
I had to update the Goodwe API as Goodwe splitted their server infrastructure per region. 
I added a server selection field in the hardware screen for that.

I also added a fix for the case where the inverter lost contact, as that might result in 0 values that mess up the database. So unless the status is "normal" we skip updating the database.

Please review the changes, as I am not a very good (C++) programmer as you may remember

Kind regards, Louis

### CODE REVIEW

File: hardware/GoodweAPI.h [hunk 1]
```diff
@@ -20,13 +21,19 @@ class GoodweAPI : public CDomoticzHardwareBase
...
 	int getSunRiseSunSetMinutes(const bool bGetSunRise);
+	bool GoodweServerClient(const std::string &sPATH, std::string &sResult);
 	float getPowerWatt(const std::string &str);
 	float getEnergyWh(const std::string &str);
 	void Do_Work();
 	void GetMeterDetails();
 	void ParseStation(const std::string &sStationId, const std::string &sStationName);
 	void ParseDeviceList(const std::string &sStationId, const std::string &sStationName);
 	void ParseDevice(Json::Value device, const std::string &sStationId, const std::string &sStationName);
+	enum _eGoodweLocation {
```
Comment (gizmocuz) [inline hardware/GoodweAPI.h]: Can you move the enum just right underneath:

```
class GoodweAPI : public CDomoticzHardwareBase
 {
```

That's more easier to find for all developers, and where i always place them ;)

File: www/views/hardware.html [hunk 2]
```diff
@@ -1662,6 +1662,15 @@ <h2><span data-i18n="Hardware">Hardware</span>:</h2>
                         <br>
                         <table class="display" id="hardwareparamsgoodweweb" border="0" cellpadding="0" cellspacing="20">
                                 <tr>
+					<td align="right" style="width:110px"><label for="comboserverselect"><span data-i18n="Select a Server">Select a Server</span>:</label></td>
```
Comment (gizmocuz) [inline www/views/hardware.html]: How about a rename from "Select a Server" to "Location", this string is already in use and translated in all/most languages

File: www/views/hardware.html [hunk 3]
```diff
@@ -1662,6 +1662,15 @@ <h2><span data-i18n="Hardware">Hardware</span>:</h2>
...
+					<td><select id="comboserverselect" style="width:200px" class="combobox ui-corner-all">
+						<option value="0">Global</option>
+						<option value="1">Oceania</option>
+						<option value="2">Europe</option>
+						</select>
+					</td>
+				</tr>
+				<tr>
 					<td align="right" style="width:110px"><label><span data-i18n="username">Goodwe-power Username</span>:</label></td>
```
Comment (gizmocuz) [inline www/views/hardware.html]: Could you change data-i18n="username" to use "Username" (capital U), this one is already in use and translated.
Also >Goodwe-power Username rename to 'Username' only, like the data-i18n, on better, you can remove that

Comment (gizmocuz) [reply www/views/hardware.html]: Best to open 'a' language file from www/i18n folder, and search what you can use

**Comment by gizmocuz** [comment]
Patch is looking perfect, i just have some tiny remarks...

**Comment by llagendijk** [comment]
HI
I made the requested changes. Thanks for the hints. I removed the duplication of the Location as well as the duplication on Username for consistency

**Comment by gizmocuz** [comment]
Thanks, can you move the enum to the cpp file ?

**Comment by llagendijk** [comment]
no problem, done

**Comment by gizmocuz** [comment]
Thanks !!

**Pull request merged**